ambry-store/src/test/java/com.github.ambry.store/HardDeleterTest.java

```diff
@@ -378,29 +378,33 @@ public void testHardDelete() {
         }
         assertEquals("Recovery range size mismatch", 0, index.hardDeleter.getHardDeleteRecoveryRange().getSize());
 
-        // indexes: [1 2] [3 4] [3d 5] [6 7] [2d 6d] [8 9*] [1d 10*] [8d 10d] [1u 3u]
-        // journal:                                                     [10d 8d 3u 1u]
+        // indexes: [1 2] [3 4] [3d 5] [6 7] [2d 6d] [8 9*] [1d 10*] [1u 3u] [8d 10d]
+        // journal:                                                    [3u 1u 10d 8d]
         tokenMovedForward = index.hardDelete();
         Assert.assertTrue(tokenMovedForward);
-        hardDeletedKeys.add(keys.get(9)); // blobId10
-        hardDeletedKeys.add(keys.get(7)); // blobId08
         if (!recordContentChecked) {
           checkRecordHardDeleted(keys, hardDeletedKeys);
         }
-        // blobid08 and blobid10
-        assertEquals("Recovery range size mismatch", 2, index.hardDeleter.getHardDeleteRecoveryRange().getSize());
 
-        // indexes: [1 2] [3 4] [3d 5] [6 7] [2d 6d] [8 9*] [1d 10] [8d 10d] [1u 3u]
-        // journal:                                                     [10d 8d* 3u 1u]
+        // indexes: [1 2] [3 4] [3d 5] [6 7] [2d 6d] [8 9*] [1d 10] [1u 3u] [8d 10d]
+        // journal:                                                    [3u 1u* 10d 8d]
         tokenMovedForward = index.hardDelete();
         Assert.assertTrue(tokenMovedForward);
+        hardDeletedKeys.add(keys.get(9)); // blobId10
+        hardDeletedKeys.add(keys.get(7)); // blobId08
         if (!recordContentChecked) {
           checkRecordHardDeleted(keys, hardDeletedKeys);
         }
+        // blobid08 and blobid10
+        assertEquals("Recovery range size mismatch", 2, index.hardDeleter.getHardDeleteRecoveryRange().getSize());
+        // Persist safeToken here since now we have two items in the recovery range and their startToken is journal-based.
+        token = (StoreFindToken) index.hardDeleter.getStartTokenSafeToPersistTo();
+        index.hardDeleter.getHardDeleteRecoveryRange().pruneTill(token);
+        // After prune we still have two
         assertEquals("Recovery range size mismatch", 2, index.hardDeleter.getHardDeleteRecoveryRange().getSize());
 
-        // indexes: [1 2] [3 4] [3d 5] [6 7] [2d 6d] [8 9*] [1d 10] [8d 10d] [1u 3u]
-        // journal:                                                     [10d 8d 3u 1u*]
+        // indexes: [1 2] [3 4] [3d 5] [6 7] [2d 6d] [8 9*] [1d 10] [1u 3u] [8d 10d]
+        // journal:                                                    [3u 1u 10d 8d*]
         // All caught up, hardDelete should not move token forward.
         tokenMovedForward = index.hardDelete();
         Assert.assertFalse(tokenMovedForward);
```

Comment:
This test is good, after verifying HardDelete doesn't move token forward, can we also verify calling `pruneHardDeleteRecoveryRange` should clear `hardDeleteRecoveryRange`?

Reply:
This is already done. Below I call persistAndAdvanceStartTokenSafeToPersist so trigger the prune and verify that the hardDeleteRecoveryRange is cleared.